config/src/lib.rs

```diff
@@ -0,0 +1,808 @@
+/// up the fs via `Remapping::find_many`.
+struct RemappingsProvider {
+    lib_paths: Vec<PathBuf>,
+    /// the root path used to turn an absolute `Remapping`, as we're getting it from
+    /// `Remapping::find_many` into a relative one.
+    root: PathBuf,
+    /// This contains either:
+    ///   - previously set remappings
+    ///   - a `MissingField` error, which means previous provider didn't set the "remappings" field
+    ///   - other error, like formatting
+    remappings: Result<Vec<Remapping>, figment::Error>,
+}
+
+impl RemappingsProvider {
+    fn new(figment: &Figment, libs: Vec<PathBuf>) -> Self {
+        let remappings = figment.extract_inner::<Vec<Remapping>>("remappings");
+        let lib_paths = figment.extract_inner::<Vec<PathBuf>>("libs").unwrap_or(libs);
+        let root = Config::find_config_file()
+            .and_then(|f| f.parent().map(|p| p.to_path_buf()))
+            .unwrap_or_else(|| std::env::current_dir().unwrap());
+        RemappingsProvider { lib_paths, root, remappings }
+    }
+}
+
+impl Provider for RemappingsProvider {
+    fn metadata(&self) -> Metadata {
+        Metadata::named("Remapping Provider")
+    }
+
+    fn data(&self) -> Result<Map<Profile, Dict>, Error> {
+        let remappings = match &self.remappings {
+            Ok(remappings) => remappings.clone(),
+            Err(err) => {
+                if let figment::error::Kind::MissingField(_) = err.kind {
+                    // only search for the remappings if weren't set before
+                    self.lib_paths
+                        .iter()
+                        .map(|lib| self.root.join(lib))
+                        .flat_map(Remapping::find_many)
+                        .collect()
+                } else {
+                    return Err(err.clone())
+                }
+            }
+        };
+
+        // turn the absolute remapping into a relative one by stripping the `root`
+        let remappings = remappings
+            .into_iter()
+            .map(|r| RelativeRemapping::new(r, &self.root).to_string())
+            .collect::<Vec<_>>();
+
+        Ok(Map::from([(
+            Config::selected_profile(),
+            Dict::from([("remappings".to_string(), figment::value::Value::from(remappings))]),
+        )]))
+    }
+
+    fn profile(&self) -> Option<Profile> {
+        Some(Config::selected_profile())
+    }
+}
+
+/// The path part of the [`Remapping`] that knows the path of the file it was configured in, if any.
+///
+/// A [`Remapping`] is intended to be absolute, but paths in configuration files are often desired
+/// to be relative to the configuration file itself. For example, a path of
+/// `weird-erc20/=lib/weird-erc20/src/` configured in a file `/var/foundry.toml` might be desired to
+/// resolve as a `weird-erc20/=/var/lib/weird-erc20/src/` remapping.
+#[derive(Debug, Clone, PartialEq)]
+pub struct RelativeRemappingPathBuf {
+    parent: Option<PathBuf>,
+    path: PathBuf,
+}
```

Comment:
Should this logic be merged with the `Remapping` struct?

Reply:
+1 on moving this to ethers-solc